File: graph/graph-bfs/go/directedgraph/directedgraph.go
package directedgraph

import (
	"container/list"
	"strconv"
	"strings"
)

// An edge in a directed graph, from Node1 to Node2.
type Edge struct {
	Node1 int
	Node2 int
}

// A directed graph, represented by adjacency lists.
type DirectedGraph struct {
	// The graph is a dictionary mapping node id (int) to a list (slice) of node id integers.
	Graph map[int][]int
}

// Create the directed graph from a list (slice) of edges (node pairs, from node1 to node2).
func (this *DirectedGraph) Create(edgelist []Edge) {
	this.Graph = make(map[int][]int, len(edgelist))
	// We'll collect the child/destination nodes in a set (children), as a map from int to "nothing".
	var nothing struct{}
	children := make(map[int]struct{})
	for k := 0; k < len(edgelist); k++ {
		edge := edgelist[k]
		this.Graph[edge.Node1] = append(this.Graph[edge.Node1], edge.Node2)
		children[edge.Node2] = nothing // Remember what all the children are.
	}

	// A node that is reachable, but has no children, will never appear as Node1,
	// so will never be key in Graph, unless we make sure to include it, as here.
	// So, add a key in the graph map for all the nodes that do not have children.
	// children has all nodes found as destinations, which also might NOT be origins.
	// For each destination node, if it is not also an origin, then add it to our graph map.
	for node, _ := range children {
		_, exists := this.Graph[node]
		if !exists {
			this.Graph[node] = make([]int, 0)
		}
	}
}

// Dump a string representation of the graph, for diagnostic display.
func (this *DirectedGraph) Dump() string {
	s := ""
	for i := 0; i < len(this.Graph); i++ {
		s += strconv.Itoa(i) + ": "
		for _, n := range this.Graph[i] {
			s += strconv.Itoa(n) + " "
		}
		s = strings.TrimRight(s, " ") + "\n"
	}
	return s
}

// Searches for the shortest path from node start to node end in breadth-first order.
// If no end is specified, then the return value is the full traversal of the graph.
// end: -1 indicates no end; traverse the whole graph.
// Returns list of the nodes traversed in the path found.
func (this *DirectedGraph) BreadthFirstSearch(start, end int) []int {
	traversal := []int{} // Resulting list of nodes visited, in order reached.

	// Remember which nodes we have visited, using a separate data structure.
	visited := make(map[int]bool)

	// The queue of the upcoming nodes to explore (just visited).
	search_from_queue := list.New()
	search_from_queue.PushBack(start)
	visited[start] = true
	traversal = append(traversal, start)

	for search_from_queue.Len() > 0 {
		e := search_from_queue.Front()
		search_from_queue.Remove(e)
		node := e.Value.(int) // Type assertion to cast the element to int.

		for _, child := range this.Graph[node] {
			if visited[child] == false {
				search_from_queue.PushBack(child)
				visited[child] = true
				traversal = append(traversal, child)
				if end != -1 && child == end {
					return traversal
				}
			}
		}
	}

	return traversal
}

File: graph/graph-bfs/typescript/DirectedGraph.ts

// We could use Record<number, Array<number>> as the graph, but
// it would nice to allow other types to identify nodes (letters, strings, etc.).
// And we would like the value returned, if the key is not there, to be undefined.
// Cf. https://brainlessdeveloper.com/typescript/2019/01/30/typescript-series-1-record-is-usually-not-the-best-choice/ .
type NodeId = string | number;
type Graph = { [P in NodeId]?: Array<NodeId> }
type Dictionary<K extends keyof any, T> = { [P in K]?: T };

// Represents a directed graph, using a node adjacency list.
// K: The type of the object identifying a node (e.g., number, string), which also is the key into the dictionary.
export default class DirectedGraph {
    // Represents the graph as a map of each node to its list of adjacent nodes (adjacency lists).
    graph: Graph;

    constructor() {
    }

    // Creates the graph from a list of edges (node pairs).
    create(edgeList: Array<[NodeId, NodeId]>) {
        this.graph = {};

        for (let edge of edgeList) {
            // Append the child to the node's growing list of children.
            if (this.graph[edge[0]]) {
                this.graph[edge[0]].push(edge[1]);
            }
            else {
                this.graph[edge[0]] = [ edge[1] ];
            }
            // A node that is reachable, but has no children, will never appear as node1, so will never be key in self.graph, unless we make sure to include it, as here.
            // Add the child as a key, with no destination yet, in case this child never has edges/children from it.
            if (this.graph[edge[1]] == null) {
                this.graph[edge[1]] = [];
            }
            //console.log(`${edge[0]}: ${this.graph[edge[0]]}`);
        }

        // Sort the children of each node, just for standardization.
        for (let node in this.graph) {
            this.graph[node].sort();
        }
    }

    // Create a string representation of the graph, for diagnostic display.
    dump() {
        let s: string = "";
        for (let node in this.graph) {
            s += `${node}: ${this.graph[node].join(" ")}\n`;
        }
        return s;
    }
    
    // Searches for the shortest path from node start to node end in breadth-first order.
    // If no end is specified, then the return value is the full traversal of the graph.
    // Returns list of the nodes traversed in the path found.
    breadth_first_search(start: NodeId, end: NodeId): Array<NodeId> {
        let traversal: Array<NodeId>= [];

        // Remember if we reached a node, using a dictionary, which is more flexible.
        let visited: Dictionary<NodeId, boolean> = {}

        let to_visit: NodeId[] = [];
        to_visit.push(start);
        visited[start] = true;
        traversal.push(start);

        while (to_visit.length > 0) {
            let node = to_visit.shift();

            for (let child of this.graph[node]) {  // Note: of, not in.
                // console.log(`${node} - ${child} - ${this.graph[node]} - ${to_visit}`);
                if (visited[child] == null) {
                    to_visit.push(child);
                    visited[child] = true;
                    traversal.push(child);
                    // Check if we reached the target and return if so.
                    if (end != null && child == end) {
                        return traversal;
                    }
                }
            }
        }

        return traversal;
    }

}


// let g = new DirectedGraph();
// let data: Array<[number, number]> = [[0, 1], [1, 2], [2, 3], [0, 0], [2, 0], [3, 1], [1, 4], [3, 4], [4, 0], [1, 5]];
// g.create(data);
// console.log(g.dump());
// let traversal = g.breadth_first_search(2, null);

// let s = `${traversal.join(" ")}`;
// console.log(s);
